[JIT] add constant prop for immutable types

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30877 [JIT] handle reassignment to inf and nan
* **#30544 [JIT] add constant prop for immutable types**
* #30543 rename shouldAnnotate api

Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.


Edit: if we were okay with running constant propagation on graphs constructed from tracing (potentially making them hard to debug), an alternative would be to run constant propagation until the graph reaches a fixed point.

Differential Revision: [D18833607](https://our.internmc.facebook.com/intern/diff/D18833607)

Commit:
add constant prop for non-aliasing types

[ghstack-poisoned]

Commit:
Update on "[JIT] add constant prop for non-aliasing types"


Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.



[ghstack-poisoned]

Commit:
Update on "[JIT] add constant prop for immutable types"


Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.



[ghstack-poisoned]

eellison commented:
Can i get a review plz

ZolotukhinM commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -3402,6 +3403,8 @@ void runCleanupPasses(std::shared_ptr<Graph>& to_clean) {
   // remove any uses of tuples that we inserted that are not needed
   LowerSimpleTuples(to_clean);
   ConstantPooling(to_clean);
+  // run constant propagation only on non-mutable types so we do not jitter
```
Nit: the comment reads as if we run constant prop to avoid jitter, but I believe that you meant that we're running const-prop **only on non-mutable types** in order for const-prop to not introduce jitter. I hope I managed to explain what I meant :)

ZolotukhinM reviewed (approved):
Looks good!

eellison commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -353,17 +388,25 @@ struct ConstantPropagator {
   GRAPH_DUMP("After ConstantPropagation: ", graph);
 }
+
+void ConstantPropagationImmutableTypes(std::shared_ptr<Graph>& graph) {
```
I could rename it primitive types as well. I think immutable is accurate though.

ZolotukhinM replied:
Up to you, just wanted to double-check.

Commit:
Update on "[JIT] add constant prop for immutable types"


Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.


Edit: if we were okay with running constant propagation on graphs constructed from tracing (potentially making them hard to debug), an alternative would be to run constant propagation until the graph reaches a fixed point.




[ghstack-poisoned]

Commit:
Update on "[JIT] add constant prop for immutable types"


Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.


Edit: if we were okay with running constant propagation on graphs constructed from tracing (potentially making them hard to debug), an alternative would be to run constant propagation until the graph reaches a fixed point.




[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -55,14 +56,31 @@ std::unordered_set<Symbol> tuple_ops = {
+    return ConstantPropagator(graph, true);
+  }
+
+  // Runs constant propagation only on ops with non-aliasing inputs
```
this is a bit confusing as a comment, maybe: only on ops that clearly do not have aliased inputs even without computing aliasing information.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -3402,6 +3402,9 @@ void runCleanupPasses(std::shared_ptr<Graph>& to_clean) {
   LowerSimpleTuples(to_clean);
   ConstantPooling(to_clean);
+  // run constant propagation on immutable types only instead of full constant
+  // prop so we do not jitter
```
"so we do not jitter" doesn't really have enough explanation here to understand it. It will be really easy to cargo-cult on this comment later (e.g. also feel like this pass is needed long after it really was because this comment doesn't describe what jitters).  Why does constant prop cause jitter when done completely but not when done to immutable things? Is it only because we haven't created a complicated enough test that only works with types like integers and tuples?

zdevito reviewed (approved):
Looks good. Some things to address in comments in line.

Commit:
Update on "[JIT] add constant prop for immutable types"


Run Constant Propagation upon compilation only on ops with non-aliasing inputs and outputs. This speeds up the first run of `torchvision.models.resnet18` by over 50% and speeds up compilation by about 25% (although the effects didn't seem additive with with https://github.com/pytorch/pytorch/pull/30503, so I'm going to land this PR first and then see if caching still has a sizable impact).

Running constant prop only with non-aliasing types does a lot of graph cleanup by removing constant ifs and a bunch of other smaller ops. It also avoids all the jitter problems we had when we tried running full constant prop previously. Bc it is idempotent it doesn't jitter, and it doesn't jitter graphs constructed from tracing because tracing doesn't emit any ops that only involve non-aliasing inputs. 

Full constant prop isn't idempotent because what ops are run depends on the state of mutation in alias db, which will often change upon successive iterations of constant propagation, and bc it affects graphs constructed from tracing.


Edit: if we were okay with running constant propagation on graphs constructed from tracing (potentially making them hard to debug), an alternative would be to run constant propagation until the graph reaches a fixed point.




[ghstack-poisoned]
